src/Http/Headers/src/ContentDispositionHeaderValue.cs

```diff
@@ -532,19 +532,11 @@ private bool RequiresEncoding(StringSegment input)
         }
 
         // Encode using MIME encoding
-        private unsafe string EncodeMime(StringSegment input)
+        private string EncodeMime(StringSegment input)
         {
-            fixed (char* chars = input.Buffer)
-            {
-                var byteCount = Encoding.UTF8.GetByteCount(chars + input.Offset, input.Length);
-                var buffer = new byte[byteCount];
-                fixed (byte* bytes = buffer)
-                {
-                    Encoding.UTF8.GetBytes(chars + input.Offset, input.Length, bytes, byteCount);
-                }
-                var encodedName = Convert.ToBase64String(buffer);
-                return "=?utf-8?B?" + encodedName + "?=";
-            }
+            var buffer = Encoding.UTF8.GetBytes(new ReadOnlySequence<char>(input.AsMemory()));
```

Comment:
Don't create a ReadOnlySequence, that's the wrong overload 😄

Reply:
There is no `Span<char>` or `ReadOnlySpan<char>` that I could see.
The closest is `int GetBytes(ReadOnlySpan<char> chars, Span<byte> bytes);`

Reply:
I see, yes, we don't want to allocate a buffer here 😄 . This is the missing overload of `Convert.ToBase64String` that we need! Conversion from utf16 to utf16:

```C#
Convert.ToBase64String(Span<char> input, Span<char> output);
```

@GrabYourPitchforks thoughts? I'd like to get an API added for this on Convert wdyt?

Reply:
@BrennanConroy go ahead with what you have for now. I'll file an issue on dotnet/runtime.

Reply:
What about using https://docs.microsoft.com/en-us/dotnet/api/system.buffers.text.base64?

You'd need to encode string into UTF8 bytes using `stackalloc byte[]` first. You could then encode in place, set nessessary bytes, and then turn the byte[] back into a string.

Reduce this method down to one string allocation. I don't know how often this is called so probably not worth the complexity.

Reply:
> You'd need to encode string into UTF8 bytes using stackalloc byte[] first

Assuming it fits, but I guess it should in the 90% case? Still we can't assume that.

Reply:
```cs
var requiredLength = "=?utf-8?B?".Length +
    Base64.GetMaxEncodedToUtf8Length(Encoding.UTF8.GetByteCount(input.AsSpan())) +
    "?=".Length;

Span<byte> buffer = requiredLength <= 256
    ? (stackalloc byte[256]).Slice(0, requiredLength)
    : new byte[requiredLength];
```
😄

Reply:
>  thoughts? I'd like to get an API added for this on Convert wdyt?

@davidfowl Base64 is a conversion of arbitrary binary data (bytes) to textual representation (ASCII chars). Not quite sure what the logical meaning would be if you were to take chars as input. If the input's already in textual representation (chars), converting it to textual representation (chars) is something of a meaningless operation.

What context am I missing here?

Reply:
In this case, the input is getting converted to UTF8 bytes. So the overloads should have an `Encoding` argument.

Wanting to convert a string to bytes, and then encode the bytes as base64 doesn't seem like a common situation. Probably doesn't warrent a helper method.

Reply:
FWIW I think @JamesNK has the right idea here. Create a buffer large enough to hold the worst-case scenario, then encode straight to that buffer. I assume the typical case is only a few dozen chars? You can use `Encoding.UTF8.GetMaxByteCount(input)`, which will tell you that the worst case scenario is ~3x expansion, then figure out what the worst-case base64 expansion is.

Rent a `byte[]` to hold the UTF-8 conversion and a `char[]` to hold the base64 conversion. Then call `string.Create`, passing in your `char[]` as the state object. This should all be allocation-free.

Reply:
> @davidfowl Base64 is a conversion of arbitrary binary data (bytes) to textual representation (ASCII chars). Not quite sure what the logical meaning would be if you were to take chars as input. If the input's already in textual representation (chars), converting it to textual representation (chars) is something of a meaningless operation.

Help me understand because I'm sure I am missing something. We have a utf16 encoded string and we convert it to utf8 byte[] to then convert that byte[] back into a base64 encoded utf16 string. Why do we need that utf8 byte[] in the middle?

Reply:
OK I get it. We actually want to encode utf8 encoded text as a base64 encoded string, because that's what the header is expecting. 

Ignore me.

Reply:
There might still be an opportunity for new API here. If we provide `Convert.ToBase64Span(ROS<byte>, Span<byte>)`, that should allow you to save the intermediate `char[]` rental. Then the only rental you'd need is the intermediate `byte[]` (for the UTF-8 conversion), and you could convert from that straight into the destination string using `string.Create`. It only saves one rental and copy in the end, but eh, maybe you'd find some other applications for it as well.

Is this interesting enough to you that we should bring this to API review? If you have evidence that this is a bottleneck we can expedite it.

Reply:
@GrabYourPitchforks can you write what the end to end usage would look like here?

Reply:
> `Convert.ToBase64Span(ROS<byte>, Span<byte>)`

This is similar to the already existing [Base64.EncodeToUtf8(ReadOnlySpan<Byte>, Span<Byte>, Int32, Int32, Boolean) Method](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.text.base64.encodetoutf8?view=net-5.0#System_Buffers_Text_Base64_EncodeToUtf8_System_ReadOnlySpan_System_Byte__System_Span_System_Byte__System_Int32__System_Int32__System_Boolean_).

I think what's needed is
```diff
public static OperationStatus EncodeToUtf8(ReadOnlySpan<byte> bytes, Span<byte> utf8, out int bytesConsumed, out int bytesWritten, bool isFinalBlock = true);
+public static OperationStatus EncodeToUtf16(ReadOnlySpan<byte> bytes, Span<char> chars, out int bytesConsumed, out int charsWritten, bool isFinalBlock = true);
```
overload. So instead of writing utf8-bytes, the new overload writes utf16-chars.
Similar to what is done [here](https://github.com/gfoidl/Base64/blob/a8a977498ae45b6dde65e134f89119784213e40f/source/gfoidl.Base64/Internal/Base64EncoderImpl.cs#L124-L136).

Assuming this overload exists, the code would look like:
```c#
static unsafe string EncodeMime(StringSegment input)
{
    int utf8Length = Encoding.UTF8.GetMaxByteCount(input.Length);
    Span<byte> utf8Buffer = utf8Length <= 256
        ? (stackalloc byte[256]).Slice(0, utf8Length)
        : new byte[utf8Length];

    int utf8BytesWritten = Encoding.UTF8.GetBytes(input.AsSpan(), utf8Buffer);
    utf8Buffer = utf8Buffer[..utf8BytesWritten];

    int base64Length = Base64.GetMaxEncodedToUtf8Length(utf8BytesWritten);

    int requiredLength = "=?utf-8?B?".Length + base64Length + "?=".Length;

    fixed (byte* ptr = &MemoryMarshal.GetReference(utf8Buffer))
    {
        return string.Create(requiredLength, (Ptr: (IntPtr)ptr, Len: utf8BytesWritten), static (buffer, state) =>
        {
            "=?utf-8?B?".AsSpan().CopyTo(buffer);
            buffer = buffer.Slice("=?utf-8?B?".Length);

            ReadOnlySpan<byte> utf8Bytes = new ReadOnlySpan<byte>(state.Ptr.ToPointer(), state.Len);
            OperationStatus status = Base64.EncodeToUtf16(utf8Bytes, buffer, out int consumed, out int written);
            Debug.Assert(status == OperationStatus.Done);
            Debug.Assert(consumed == state.Len);

            buffer = buffer.Slice(written);
            "?=".AsSpan().CopyTo(buffer);
        });
    }
}
```
(BTW: is there a better way to pass the ROS to string.Create?)

Without this overload it would look like:
```c#
static unsafe string EncodeMime(StringSegment input)
{
    int utf8Length = Encoding.UTF8.GetMaxByteCount(input.Length);
    Span<byte> utf8Buffer = utf8Length <= 256
        ? (stackalloc byte[256]).Slice(0, utf8Length)
        : new byte[utf8Length];

    int utf8BytesWritten = Encoding.UTF8.GetBytes(input.AsSpan(), utf8Buffer);
    utf8Buffer = utf8Buffer[..utf8BytesWritten];

    int base64Length = Base64.GetMaxEncodedToUtf8Length(utf8BytesWritten);

    int requiredLength = "=?utf-8?B?".Length + base64Length + "?=".Length;

    fixed (byte* ptr = &MemoryMarshal.GetReference(utf8Buffer))
    {
        return string.Create(requiredLength, (Ptr: (IntPtr)ptr, Len: utf8BytesWritten, Base64Len: base64Length), static (buffer, state) =>
        {
            "=?utf-8?B?".AsSpan().CopyTo(buffer);
            buffer = buffer.Slice("=?utf-8?B?".Length);

            Span<byte> base64Utf8Buffer = state.Base64Len <= 256
                ? (stackalloc byte[256]).Slice(0, state.Base64Len)
                : new byte[state.Base64Len];

            ReadOnlySpan<byte> utf8Bytes = new ReadOnlySpan<byte>(state.Ptr.ToPointer(), state.Len);
            OperationStatus status = Base64.EncodeToUtf8(utf8Bytes, base64Utf8Buffer, out int consumed, out int written);
            Debug.Assert(status == OperationStatus.Done);
            Debug.Assert(consumed == state.Len);
            Debug.Assert(written == base64Utf8Buffer.Length);

            written = Encoding.UTF8.GetChars(base64Utf8Buffer, buffer);

            buffer = buffer.Slice(written);
            "?=".AsSpan().CopyTo(buffer);
        });
    }
}
```
so with the additional conversion from utf8-bytes -> chars.

Reply:
> `Convert.ToBase64Span(ROS<byte>, Span<byte>)`

Having another though on this and some other use-cases. Maybe there should be 
```diff
public static class Convert
{
    public static string ToBase64String(ReadOnlySpan<byte> bytes, Base64FormattingOptions options = Base64FormattingOptions.None);
+   public static string ToBase64String(ReadOnlySpan<byte> bytes, ReadOnlySpan<char> preBuffer, ReadOnlySpan<char> postBuffer, Base64FormattingOptions options = Base64FormattingOptions.None);
}
```
so it's usage can be generalized, and the method here could be written as simply as
```c#
static unsafe string EncodeMime(StringSegment input)
{
    int utf8Length = Encoding.UTF8.GetMaxByteCount(input.Length);
    Span<byte> utf8Buffer = utf8Length <= 256
        ? (stackalloc byte[256]).Slice(0, utf8Length)
        : new byte[utf8Length];

    int utf8BytesWritten = Encoding.UTF8.GetBytes(input.AsSpan(), utf8Buffer);
    utf8Buffer = utf8Buffer[..utf8BytesWritten];

    return Convert.ToBase64String(utf8Buffer, "=?utf-8?B?", "?=");
}
```

Reply:
You don't need an `OperationStatus`-based API or any pre-buffer or post-buffer for this proposal. The pseudo-code would be:

```cs
int maxUtf8Length = Encoding.UTF8.GetMaxByteCount(inputString);
byte[] inputAsUtf8 = Rent<byte>(maxUtf8Length);
int actualUtf8Length = Encoding.UTF8.GetBytes(inputString, inputAsUtf8);
int requiredBase64CharCount = CalculateRequiredBase64Length(actualUtf8Length);

string retVal = string.Create("prefix".Length + "suffix".Length + requiredBase64CharCount, (inputAsUtf8, actualUtf8Length) => (span, state)
{
    "prefix".AsSpan().CopyTo(span);
    span = span.Slice("prefix".Length);
    int actualBase64CharsWritten = Convert.ToBase64Span(inputAsUtf8.Slice(0, actualUtf8Length), span);
    span = span.Slice(actualBase64CharsWritten);
    "suffix".AsSpan().CopyTo(span);
    span = span.Slice("suffix".Length);
    if (!span.IsEmpty) { FailFast("?????"); }
});

Return(inputAsUtf8);
return retVal;
```

Reply:
Isn't the problem with that code that you need the **exact** base64 length, otherwise you'll hit the `if (!span.IsEmpty) { FailFast("?????"); }` path sometimes.

I've gone with James' pseudo code as it avoids the byte[] allocation in the common case.

Reply:
> Isn't the problem with that code that you need the exact base64 length?

Yes, but this can be exactly calculated from the input `byte[]` length. Here's how the BCL does it. It's only two lines plus an overflow check.

https://github.com/dotnet/runtime/blob/82ca681cbac89d813a3ce397e0c665e6c051ed67/src/libraries/System.Private.CoreLib/src/System/Convert.cs#L2530-L2531

Feel free to copy and paste that logic. We can also make it into its own public API if needed.

Reply:
https://github.com/dotnet/runtime/blob/82ca681cbac89d813a3ce397e0c665e6c051ed67/src/libraries/System.Memory/src/System/Buffers/Text/Base64Encoder.cs#L155 that's the same and simpler calculation. Or use that method, it's already public.

> You don't need an `OperationStatus`-based API or any pre-buffer or post-buffer for this proposal.

Right. But this needs a new `Convert.ToBase64Span` method. 
My concern is only about having two base64-API sets. One in Convert-class and one in System.Buffers.Text.Base64. I'd prefer the latter, as it's name clearly indicates it is about base64.

> `Base64.EncodeToUtf8InPlace`

@BrennanConroy just to note this method is not vectorized, whilst `Base64.EncodeToUtf8` is for size >= 16. So depending on the measure here it may be more advantageous (see second snippet above).

Reply:
For some reason the following code is slower and supposedly allocates more than the current impl (using BenchmarkDotnet):
```
private static ReadOnlySpan<char> MimePrefix => new char[] { '"', '=', '?', 'u', 't', 'f', '-', '8', '?', 'B', '?' };
private static ReadOnlySpan<char> MimeSuffix => new char[] { '?', '=', '"' };
// ...

var utf8Length = Encoding.UTF8.GetByteCount(input.AsSpan());
var buffer = ArrayPool<byte>.Shared.Rent(utf8Length);
try
{
	Encoding.UTF8.GetBytes(input.AsSpan(), buffer);

	return string.Create(MimePrefix.Length + MimeSuffix.Length + GetBase64Length(utf8Length), (buffer, utf8Length), static (span, state) =>
	{
		MimePrefix.CopyTo(span);
		span = span.Slice(MimePrefix.Length);
		var ret = Convert.TryToBase64Chars(state.buffer.AsSpan().Slice(0, state.utf8Length),
			span, out var written);
		Debug.Assert(ret);

		span = span.Slice(written);
		MimeSuffix.CopyTo(span);

		if (span.Length != MimeSuffix.Length)
		{
			throw new InvalidOperationException();
		}
	});
}
finally
{
	ArrayPool<byte>.Shared.Return(buffer);
}
```
And test:
```
var contentDisposition = new ContentDispositionHeaderValue("inline");
contentDisposition.FileName = "FileÃName.bat";
```

Comment above:
| Method |     Mean |    Error |   StdDev |        Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------- |---------:|---------:|---------:|------------:|-------:|------:|------:|----------:|
|      CD | 977.1 ns | 32.89 ns | 95.43 ns | 1,023,416.6 | 0.0156 |     - |     - |     552 B |

Current:
| Method |     Mean |    Error |   StdDev |        Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------- |---------:|---------:|---------:|------------:|-------:|------:|------:|----------:|
|      CD | 790.1 ns | 16.70 ns | 46.56 ns | 1,265,685.7 | 0.0078 |     - |     - |     312 B |

I'm going to stop work on this and merge as is assuming there are no glaring issues with the recent changes during PR review.

Reply:
> For some reason the following code is slower and supposedly allocates more than the current impl (using BenchmarkDotnet):

The Roslyn optimization that embeds arrays of consts exposed as ReadOnlySpans into the data section, thereby avoiding a per-access array allocation, only applies when the array is a byte[], sbyte[], or bool[], i.e. when it's an array of byte-sized types.  Otherwise, there are endianness concerns.  https://github.com/dotnet/runtime/issues/24961 would be needed and then Roslyn respecting it before it could be used with other types, like the char[] in your repro.

Reply:
Makes sense, didn't think about that optimization only applying to `ReadOnlySpan<byte>`

Reply:
Yeah, it's subtle.